Fixes Search that would return no result if semicolon is used

venditdevs opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | Fixes searching multiple products seperated by semicolon. Also changed default of $eligibleProducts2 to array so an unnecessary if/else could be removed.
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #25394.
| How to test?      | Try to search multiple products separated by a semicolon based on tags. Example of searchs<br/>- "red shirt"<br/>- "red"<br/>- "red shirt ; blue shirt"<br/>- "red ; blue"<br/>The Search system will return the results for "red" and for "blue" rather than trying to find results that match simultaneously "red" and "blue".
| Possible impacts? | Search engine, more information in the issue #25394


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/25396)
<!-- Reviewable:end -->

venditdevs committed:
Search with semicolon gives no result.

Fixes this issue:
https://github.com/PrestaShop/PrestaShop/issues/25394

intersect_array isn't the right function in this foreach.

Also changed default of $eligibleProducts2 to array so an if/else later on could be removed

Progi1984 commented:
@venditdevs Hi Could you fix PHPCSFixer Check ?

matks commented on classes/Search.php:
```diff
@@ -319,21 +319,18 @@ public static function find(
-                $eligibleProducts2 = array_intersect($eligibleProducts2, $productIds);
-            }
-
+            $eligibleProducts2 = array_merge($eligibleProducts2, $productIds);
```
This is a big change of behavior.
Previously we would look at the intersection of `$eligibleProducts2` and `$productIds` now we merge them. I expect that this will make the search less accurate, it will present way much results than before.

I am not sure this is the right way to fix the issue

> It's this line: $eligibleProducts2 = array_intersect($eligibleProducts2, $productIds);
This will return the differences between the two arrays which is none when you use 2 or more search words and so the result is empty.

Can you tell us
- what is inside `$eligibleProducts2` when a single search word is used
- what is inside `$eligibleProducts2` when two search words are used

I think the root issue might be there.

venditdevs replied:
@matks : If the two arrays don't match.....the array $eligibleProducts2 is empty. So if word 1 result is (e.g.) id_product 5 and word 2 result is id_product 6. You have NO results because they don't match. That's wrong in every way you look at it.

Data:
testproduct1 = id_product 1
testproduct2 = id_product 2
testproduct3 = id_product 3

What is inside $eligibleProducts2 when a single search word is used:
Example: Search for testproduct1; the contents of $eligibleProducts2 = array([0]=>1)

What is inside $eligibleProducts2 when two search words are used:
Example: Search for testproduct1;testproduct2; the contents of $eligibleProducts2 = array()
Reason: The 2 arrays don't match and so array_intersect returns an empty array. But it really should return array([0]=>1,[1]=>2)

What is inside $eligibleProducts2 when three search words are used:
Example: Search for testproduct1;testproduct2;testproduct3; the contents of $eligibleProducts2 = array()
Reason: The 2 arrays don't match and so array_intersect returns an empty array. But it really should return array([0]=>1,[1]=>2,[2]=>3)

matks replied:
Aaaah I got it now.

I think that the goal of `$eligibleProducts2 = array_intersect($eligibleProducts2, $productIds);` was to make search more accurate.

Exemple: I search "red shirt".
First loop on "red" will give me 4 red products
Second loop on "shirt" will give me 5 shirt products
Then intersection will give me the red shirt products

If we remove the `array_intersect()` we will get every red product and every shirt product. We don't want that to happen.

But if you search, like you do, the words "shirt ; trousers" you want to find both shirts and trousers
but the above logic will find all products that are shirts AND trousers and this is not possible.

So the main issue is that the current logic attempts to find "products that match all input words" while in your usecase you want "products that match any input word".

So I cannot approve your change. We need to find a way that preserve both behavior, if that's possible.

venditdevs replied:
Maybe with a strpos check on ; set a flag automatically to search for all words. So if ; is present in $expr search for all words.
Would that be a fast and easy fix?

matks replied:
> Maybe with a strpos check on ; set a flag automatically to search for all words. So if ; is present in $expr search for all words.

That could work for your usecase, but what about "red shirt ; blue trousers ; black shoes" 😅 

I guess the very best would be
1. When we detect there are terms separated by `;`
2. We select each part separated by `;`
3. For each of these parts, we perform the regular intersect search : that means we look for red shirts, then we look for blue trousers, then we look for black shoes
4. Finally we merge all results together, so we return results that are red shirts OR blue trousers OR black shoes

It's a more complex though. Also such a complex behavior cannot be guessed, we will need
- to test it
- to document it

matks replied:
At this point it's probably a good idea to ask @PrestaShop/prestashop-core-developers for their opinion

venditdevs replied:
"That could work for your usecase, but what about "red shirt ; blue trousers ; black shoes""
I think this can be fixed quite easy ..if we do an explode on ';' in $expr build a foreach. 
At the end tie everything together.

This would fix everything imho.

I'm going to test it

venditdevs replied:
@matks  . I tested it. If we explode the expressions on semicolon. Loop it in a for each it will give the expected result. In all ways

venditdevs replied:
Tested it with code the find() function of Prestashop 1.7.7.5. However it won't work with the code in 'Develop' branche. It will mess up the result later on in the code. Working on that. The current outcome: $eligibleProducts2Full (new var) has 2 product id's however only 1 product is returned as final result.

venditdevs replied:
PR updated!

venditdevs committed:
Updated as in discussion

venditdevs commented:
> @venditdevs Hi Could you fix PHPCSFixer Check ?

I'm sorry. I really have no idea what the problem in PHPCSFixer is and how to fix.

matks commented:
> > @venditdevs Hi Could you fix PHPCSFixer Check ?
> 
> I'm sorry. I really have no idea what the problem in PHPCSFixer is and how to fix.

You can run it locally
```
php /vendor/bin/php-cs-fixer fix
```

venditdevs committed:
CSFixer fix

venditdevs commented:
> > > @venditdevs Hi Could you fix PHPCSFixer Check ?
> > 
> > 
> > I'm sorry. I really have no idea what the problem in PHPCSFixer is and how to fix.
> 
> You can run it locally
> 
> ```
> php /vendor/bin/php-cs-fixer fix
> ```

OK thx! 
Pulled the Git Repo and ran cs-fixer. Edited PR ...checks are all queued...guess we'll have to wait if it's fixed now.

venditdevs commented:
@matks ...All unit tests are failing on setup :S That's nothing i can fix right?

matks commented:
> @matks ...All unit tests are failing on setup :S That's nothing i can fix right?

Indeed looks like we have a bigger issue 😆  stand by

venditdevs commented:
> > @matks ...All unit tests are failing on setup :S That's nothing i can fix right?
> 
> Indeed looks like we have a bigger issue  stand by

Well at least cs-fixer is ok now :)

venditdevs commented:
@matks . I'm testing further and there is much more broken then we thought in the routine. If you search for this:
testproduct1;testproduct2;test it won't return the first 2 product but the whole search is empty again.

It's really starting to annoy me....

venditdevs committed:
More bugfixes.

Bugfixed: 
1) Empty result if 1 expression result was 0 products but there were already products found on the other expressions
2) $wordCnt was on wrong position after my code change

Tested it again in every way. It now behaves as it should

matks commented:
> @matks . I'm testing further and there is much more broken then we thought in the routine. If you search for this:
> testproduct1;testproduct2;test it won't return the first 2 product but the whole search is empty again.

Search logic has always been a complex topic. This is why there are companies that have search as a core business like Algolia

I don't think we should aim to cover all possible searches in PrestaShop. PrestaShop relies on SQL and SQL is not a good engine for search. If someone really wants nice search results, I think that there are some nice modules (like Algolia, but also others) that will do a lot better than what we can achieve with SQL. Product search is not the main focus on PrestaShop, main focus is to sell. Let's implement something nice, simple, and that covers basic usecase :) that will already be quite good, what do you think?

venditdevs commented:
> that will already be quite good, what do you think?

@matks Well, the results are better than it used to be, that's for sure;)

I fixed the last problem in the current PR. So everything now works as it used to be in PS 1.6.6.X AND uses the new routine from 1.7.7.X that gives better search results. So: Black shoes;red shirts;green trousers will work now as expected.

matks commented:
> I fixed the last problem in the current PR. So everything now works as it used to be in PS 1.6.6.X AND uses the new routine from 1.7.7.X that gives better search results. So: Black shoes;red shirts;green trousers will work now as expected.

I think this is a nice improvement 😄  no need to go further.

matks reviewed (approved):
Code looks good to me. I approve the PR.

I add 2 things that could be great, if you are willing to go further:
- document this "loop on the items separated by a semi-colon" behavior in the function PHPDoc, to help future readers to understand what this part of the code does
- write automated tests to validate this behavior

Hlavtox commented:
@matks @venditdevs Guys, there were big argues about the search. Exactly by this. You searched "red car" and it returned everything red and every car. Completely unusable. 

Now it works, **PLEASE** don't screw it up.

matks commented:
> @matks @venditdevs Guys, there were big argues about the search. Exactly by this. You searched "red car" and it returned everything red and every car. Completely unusable.
> 
> Now it works, **PLEASE** don't screw it up.

We can't let https://github.com/PrestaShop/PrestaShop/issues/25394 unsolved. There's even a placeholder to suggest using the semicolon.

If you wan't to be sure these changes are fine, the best is to check them out and verify them 😄

matks commented:
Just found this 😛  https://github.blog/changelog/2021-08-02-search-issues-by-label-using-logical-or/ : as we can, it's standard to use comma or semicolon and expect a OR search being performed

khouloudbelguith commented:
Hi @venditdevs,

It is ok :heavy_check_mark: 

**Before** 
![image](https://user-images.githubusercontent.com/16067358/131370573-db72e7e0-45fb-46b9-b25e-396ce06cc8c9.png)
**After**
![image](https://user-images.githubusercontent.com/16067358/131370629-3381d473-914f-4d91-bcdf-96f8bd46405b.png)

**With PR**
![image](https://user-images.githubusercontent.com/16067358/131370708-c24af1ba-e677-47f8-9061-39068c2c9f0d.png)
![image](https://user-images.githubusercontent.com/16067358/131370742-2a616f09-b04e-4855-8c89-62a92e035ed8.png)

>https://github.com/PrestaShop/PrestaShop/pull/25396#issuecomment-889745908

**Without PR**
![image](https://user-images.githubusercontent.com/16067358/131370940-6891b740-4fdf-4c5e-9f0f-437bd04319ef.png)

**With PR**
![image](https://user-images.githubusercontent.com/16067358/131370895-bfbf3724-c650-4fcb-9a0e-62108bff1eae.png)

Thanks!

atomiix commented:
Thanks everyone! :)

matks commented:
> @matks @venditdevs Guys, there were big argues about the search. Exactly by this. You searched "red car" and it returned everything red and every car. Completely unusable.
> 
> Now it works, **PLEASE** don't screw it up.

@Hlavtox the current changes only handle the semicolon separator as a OR

Can you have a look and tell is result looks fine ? 😉  from my opinion it looks good
